design/healcheck-endpoint.md

```diff
@@ -0,0 +1,132 @@
+and External IdPs probes are free of charge as we report success/failure calls for them in the Metrics components.
+With this approach, the Healthcheck Endpoint becomes an alternative reporting tool for the Metrics.
+
+This design assumes implementing Health Endpoint in Passive Mode (with possible elements of active checks).
+
+#### Microprofile Health
+
+[Eclipse Microprofile Health](https://download.eclipse.org/microprofile/microprofile-health-3.0-RC4/microprofile-health-spec.html)
+project provides capabilities for reporting Liveness and Readiness probes. In the contrast to Microprofile Metrics,
+the Health project doesn't provide a centralized class to access all configured checks. The SmallRye implementation
+provides `SmallRyeHealthReporter` (see [the code](https://github.com/smallrye/smallrye-health/blob/73e9ea8c31498dab7753f4a428cb424bf9fbaa1a/implementation/src/main/java/io/smallrye/health/SmallRyeHealthReporter.java#L49))
+, that is a CDI-annotated Bean that contains references to `HealthCheck` implementations. The `HealthCheck` interface
+implementations provides a specific check - e.g. check if a database is up.
+
+Keycloak will provide either its own implementation of a class similar to `SmallRyeHealthReporter`, or
+a CDI-annotated bean that will be used as a bridge between Microprofile Health, and the rest of the Keycloak code. The
+latter solution is preferred as the out-of-the-box support has been already implemented in Quarkus.
+
+Individual Healthcheck implementations, such as database check, External IdP check or LDAP check need to be provided
+as `HealthCheck` class implementations.
+
+#### Database check
+
+Agroal (the lightweight database connection pool) publishes a set of metrics into the main Quarkus Metrics Registry
+(see [the code](https://github.com/quarkusio/quarkus/blob/master/extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/metrics/AgroalMetricsRecorder.java#L37-L113)).
+It also provides an active healthcheck (see [the code](https://github.com/quarkusio/quarkus/blob/master/extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/health/DataSourceHealthCheck.java)).
+
+The goal of the implementation is to provide a specialized check that is a hybrid between an active and a passive mode.
+When Keycloak is serving traffic, the implementation needs to track the database health by observing metrics reported
+by the Agroal framework. However, when Keycloak is idle, the implementation needs to either ask Agroal to maintain at least
+one, healthy connection or perform an active check.
+
+#### Healthcheck API for Liveness and Readiness probes
+
+The REST response has been defined in [Microprofile Health Spec](https://download.eclipse.org/microprofile/microprofile-health-3.0-RC4/microprofile-health-spec.html#_appendix_a_rest_interfaces_specifications).
+There are 3 possible HTTP Status codes that might be returned from the endpoint:
+* 200 - Healthcheck passes
+* 500, 503 - Healthcheck fails
```

Comment:
Looking at the [link from above](https://download.eclipse.org/microprofile/microprofile-health-3.0-RC4/microprofile-health-spec.html#_appendix_a_rest_interfaces_specifications), the 200 status can also provide key-value detailed report.
Could you provide sample response for all of the three status codes? What kind of `data` datapoints would be mandatorily exposed for any health check, if any? Would e.g. `cause` be mandatory for any failed check in 500 / 503 response?

Reply:
That's a good point @hmlnarik !

I believe it doesn't make sense to expose anything for a successful check. Successful checks are not logged anywhere in Kubernetes (failed are - in Events).

HTTP 503 is a standard response for a failed probe. It should contain enough debugging information to help users investigating a failure. In our case - this should probably the time that the probe fails. It might also be a good idea to expose an exception message but we need to make sure, there's no secret information with it. 

HTTP 500 indicates an exception thrown in the probe's code.

Reply:
I slightly updated the text. I hope it works for you.